torch/distributed/elastic/rendezvous/dynamic_rendezvous.py

```diff
@@ -243,6 +245,129 @@ def __init__(self) -> None:
         self.last_keep_alives = {}
 
 
+class _RendezvousStateHolder(ABC):
+    """Holds the rendezvous state synced with other nodes."""
+
+    state: _RendezvousState
+
+    def __init__(self) -> None:
+        self.state = _RendezvousState()
+
+    @abstractmethod
+    def sync(self) -> None:
+        """Gets or sets the latest rendezvous state."""
+
+    @abstractmethod
+    def mark_dirty(self) -> None:
+        """Marks the local rendezvous state as dirty."""
+
+
+class _RendezvousStateHolderImpl(_RendezvousStateHolder):
+    """Represents the default implementation of ``RendezvousStateHolder``.
+
+    Args:
+        backend:
+            The rendezvous backend to use.
+        settings:
+            The rendezvous settings.
+        cache_duration:
+            The amount of time, in seconds, to cache the last rendezvous state
+            before requesting it from the backend again.
+    """
+
+    backend: RendezvousBackend
+    settings: RendezvousSettings
+    cache_duration: int
+
+    _token: Token
+    _dirty: bool
+    _last_sync_time: float
+
+    def __init__(
+        self, backend: RendezvousBackend, settings: RendezvousSettings, cache_duration: int = 1
+    ) -> None:
+        super().__init__()
+
+        self.backend = backend
+        self.settings = settings
+        self.cache_duration = cache_duration
+
+        self._token = None
+        self._dirty = False
+        self._last_sync_time = -1
+
+    def sync(self) -> None:
+        """See base class."""
+        if self._dirty:
+            state_bits = pickle.dumps(self.state)
+
+            response = self.backend.set_state(state_bits, self._token)
```

Comment:
its a little bit hard to know how it is going to be used yet, but:

It seems if the node has a dirty state, and the store has also updated state, the node's token will be different from the one on on the store. In this case the response will contain the latest state, and it would override the current node's state. Is it the intended behavior?

Reply:
Yes, that is right. Whichever node (that has the same token along with a set of other nodes) acts first, updates the state; other nodes have to retry after fetching the latest state. This ultimately functions as a distributed CAS loop.